### PR TITLE
PERF2177 add MultiCollectionInsert for LLT Insert Workload

### PR DESCRIPTION
**Pull request by jimoleary**
MultiCollectionInsert makes this simpler and avoids most of the issues I had hit in terms of complex yml and slow startup.

I add some clion documentation w.r.t to building and running.

@rtimmons Would it be an idea to add a './run-genny cedar' command to hide the curator details (moving the metrics, running the curator command etc.). It would only be needed when running genny_core directly so it wouldn't need to be documented with  './run-genny --help'

### CODE REVIEW

File: src/cast_core/include/cast_core/actors/MultiCollectionInsert.hpp
Comment (dalyd) [inline src/cast_core/include/cast_core/actors/MultiCollectionInsert.hpp:31]: s/performs updates/performs inserts/

File: src/cast_core/include/cast_core/actors/MultiCollectionInsert.hpp
Comment (dalyd) [inline src/cast_core/include/cast_core/actors/MultiCollectionInsert.hpp:33]: s/update/insert/

File: src/cast_core/include/cast_core/actors/MultiCollectionInsert.hpp
Comment (dalyd) [inline src/cast_core/include/cast_core/actors/MultiCollectionInsert.hpp:34]: s/updated/inserted/

File: src/cast_core/include/cast_core/actors/OptionsConversion.hpp
Comment (dalyd) [inline src/cast_core/include/cast_core/actors/OptionsConversion.hpp:1]: Nice on pulling this out

Comment (dalyd) [reply src/cast_core/include/cast_core/actors/OptionsConversion.hpp:1]: Any changes besides location of the code?

File: src/cast_core/src/MultiCollectionInsert.cpp
Comment (dalyd) [inline src/cast_core/src/MultiCollectionInsert.cpp:68]: Should these three lines be called out into a function? They should be the same in a lot of places, yes?

File: src/cast_core/src/MultiCollectionInsert.cpp
Comment (dalyd) [inline src/cast_core/src/MultiCollectionInsert.cpp:73]: Why can't you just use config->batchsize below?

File: src/cast_core/src/MultiCollectionInsert.cpp
Comment (dalyd) [inline src/cast_core/src/MultiCollectionInsert.cpp:77]: Your opCtx is going to include the document generation time in the operation time. If you move it below this for loop you will only have the time for the insert_many call. Not sure if this was intentional or not. For Loader I intentionally had it after.

File: src/cast_core/test/MultiCollectionInsert_test.cpp
Comment (dalyd) [inline src/cast_core/test/MultiCollectionInsert_test.cpp:34]: Um

**Review by dalyd — COMMENTED** [review]
I focused on the actor itself, and less the workload.

File: src/cast_core/include/cast_core/actors/MultiCollectionInsert.hpp
Comment (jimoleary) [reply src/cast_core/include/cast_core/actors/MultiCollectionInsert.hpp:31]: :+1:  and all the following.

File: src/cast_core/include/cast_core/actors/MultiCollectionInsert.hpp
Comment (jimoleary) [reply src/cast_core/include/cast_core/actors/MultiCollectionInsert.hpp:33]: :+1:

File: src/cast_core/include/cast_core/actors/MultiCollectionInsert.hpp
Comment (jimoleary) [reply src/cast_core/include/cast_core/actors/MultiCollectionInsert.hpp:34]: :+1:

File: src/cast_core/include/cast_core/actors/OptionsConversion.hpp
Comment (jimoleary) [reply src/cast_core/include/cast_core/actors/OptionsConversion.hpp:1]: No this was just moving the code out. I'm not terrifically happy with the location but it will do for now.

I will cut a new ticket to refactor the other actors to use the conversion operators once thiis code is merged.

File: src/cast_core/src/MultiCollectionInsert.cpp
Comment (jimoleary) [reply src/cast_core/src/MultiCollectionInsert.cpp:68]: I think that is a good idea but also another ticket.

File: src/cast_core/src/MultiCollectionInsert.cpp
Comment (jimoleary) [reply src/cast_core/src/MultiCollectionInsert.cpp:73]: removing.

File: src/cast_core/src/MultiCollectionInsert.cpp
Comment (jimoleary) [reply src/cast_core/src/MultiCollectionInsert.cpp:77]: Good call. Is it worth tracking bytes also?

File: src/cast_core/test/MultiCollectionInsert_test.cpp
Comment (jimoleary) [reply src/cast_core/test/MultiCollectionInsert_test.cpp:34]: This was generated by the actor generation script. I'll remove the comment.

File: src/cast_core/src/MultiCollectionInsert.cpp
Comment (dalyd) [reply src/cast_core/src/MultiCollectionInsert.cpp]: Yeah, why not?

File: README.md
Comment (rtimmons) [inline README.md:77]: Annoyance but can you put these images on the "images issue" https://github.com/mongodb/genny/issues/296? Just add them as a comment and grab the URL. This saves us from having binary assets that get pulled on every clone. 

These images are small, but it'll add up (and it makes it easier to add more images if you don't have to think much about image sizes)

File: README.md
Comment (rtimmons) [inline README.md:88]: I didn't even know this was a thing.
If you cut a TIG ticket I'd be happy to incorporate this into the `run-genny` CLI. I think you'd suggested a `run-genny curator` thing that would do this and that seems 💯

File: src/cast_core/include/cast_core/actors/OptionsConversion.hpp
Comment (rtimmons) [inline src/cast_core/include/cast_core/actors/OptionsConversion.hpp:32]: for my own sanity - did you add/change these at all or just straight up moving them here? I ask because I had to change them to `genny::NodeConvert` for https://github.com/mongodb/genny/pull/424. If I know what changes there were (if any) it'll make merging in these changes a lot easier.

File: src/cast_core/src/MultiCollectionInsert.cpp
Comment (rtimmons) [inline src/cast_core/src/MultiCollectionInsert.cpp:56]: 🤷🏽 leading `_` or not

File: src/cast_core/src/MultiCollectionInsert.cpp
Comment (rtimmons) [inline src/cast_core/src/MultiCollectionInsert.cpp:79]: you can `addBytes` multiple times--it's just incrementing an internal counter of its own.

File: src/cast_core/src/MultiCollectionInsert.cpp
Comment (rtimmons) [inline src/cast_core/src/MultiCollectionInsert.cpp:85]: do you care to assert that `inserted_count == batchSize`?

File: src/cast_core/test/MultiCollectionInsert_test.cpp
Comment (rtimmons) [inline src/cast_core/test/MultiCollectionInsert_test.cpp:64]: I think this will fail? If not it..should..I think. At any rate I think for this to be correct you would need to set `CollectionCount` to some reasonable number and then do the `count_documents` over each of the possible collections.

File: src/workloads/scale/LLTInsert.yml
Comment (rtimmons) [inline src/workloads/scale/LLTInsert.yml:53]: I don't have a good context for this, but this seems..really high..like to the point of potentially over-saturating the workload client rather than the server. Idk. We don't have much documentation or guidance in place for pool size, but consider a comment justifying this number.

**Review by rtimmons — CHANGES_REQUESTED** [review]
LGTM modulo a few nits. Biggest thing is the test. Please push back wherever you disagree.

File: README.md
Comment (jimoleary) [reply README.md:77]: Will do.

File: README.md
Comment (jimoleary) [reply README.md:88]: [TIG-2934](https://jira.mongodb.org/browse/TIG-2934).

File: src/cast_core/include/cast_core/actors/OptionsConversion.hpp
Comment (jimoleary) [reply src/cast_core/include/cast_core/actors/OptionsConversion.hpp:32]: Only cut and paste, no changes. Although I do have changes in my next PR.

File: src/cast_core/src/MultiCollectionInsert.cpp
Comment (jimoleary) [reply src/cast_core/src/MultiCollectionInsert.cpp:56]: renamed to _insertOptions_ (also removed the queryExpr comment). What is the preferred convention? It seems to vary actor to actor.

File: src/cast_core/src/MultiCollectionInsert.cpp
Comment (jimoleary) [reply src/cast_core/src/MultiCollectionInsert.cpp:79]: I am doing a single add afterwards so that the vector creation is not part of the operation (that may be over kill).

File: src/cast_core/src/MultiCollectionInsert.cpp
Comment (jimoleary) [reply src/cast_core/src/MultiCollectionInsert.cpp:85]: I do not (for now, also none of the other actors seem to assert in this fashion). But I have added a check to ensure that result is available.

File: src/cast_core/test/MultiCollectionInsert_test.cpp
Comment (jimoleary) [reply src/cast_core/test/MultiCollectionInsert_test.cpp:64]: Is the test not being run as part of the commit? I'll update to the test now anyway.

**Review by dalyd — APPROVED** [review]
I'll get out the way. Looks like good stuff.

File: src/workloads/scale/LLTInsert.yml
Comment (jimoleary) [reply src/workloads/scale/LLTInsert.yml:53]: Not sure how I missed this comment. I was messing around with cases where a large pool size was required but 2000 is too high.

**Comment by jimoleary** [comment]
@rtimmons any more feedback on this one?

File: src/workloads/docs/MultiCollectionInsert.yml
Comment (rtimmons) [inline src/workloads/docs/MultiCollectionInsert.yml:19]: I think the above comments are still funky, plus I think an example including `OperationOptions` would be helpful.

**Review by rtimmons — APPROVED** [review]

File: src/workloads/docs/MultiCollectionInsert.yml
Comment (jimoleary) [reply src/workloads/docs/MultiCollectionInsert.yml:19]: I have another pr lined up, I'll fix the example in that.

Thanks.

**Pull request merged**
